src/resolve/resolved-dns-stream.c

```diff
@@ -441,6 +453,34 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
                         log_warning_errno(errno, "Couldn't restart TCP connection timeout, ignoring: %m");
         }
 
+#if ENABLE_DNS_OVER_TLS
+        if (s->encrypted) {
+                uint32_t events;
+
+                r = sd_event_source_get_io_events(s->io_event_source, &events);
+                if (r < 0)
+                        return r;
+
+                /* When using DNS-over-TLS, the underlying TLS library may read the entire TLS record
+                   and buffer it internally. If this happens, we will not receive further EPOLLIN events,
+                   and unless there's some unrelated activity on the socket, we will hang until time out.
+                   To avoid this, if there's buffered TLS data, generate a "fake" EPOLLIN event.
+                   This is hacky, but it makes this case transparent to the rest of the IO code. */
+                if ((events & EPOLLIN) != 0 && dnstls_stream_has_buffered_data(s)) {
```

Comment:
Not a matter, but I prefer `FLAGS_SET(events, EPOLLIN)`.

Reply:
Agreed and done.